From: Dan Williams <dan.j.williams@intel.com>
Date: Wed, 6 Jan 2016 12:03:41 -0800
Subject: block: clarify badblocks lifetime
Git-commit: 20a308f09e0d29ce6f5a4114cc476a998d569bfb
Patch-mainline: v4.5-rc1
References: bsc#984203

The badblocks list attached to a gendisk is allocated by the driver
which equates to the driver owning the lifetime of the object.  Do not
automatically free it in del_gendisk(). This is in preparation for
expanding the use of badblocks in libnvdimm drivers and introducing
devm_init_badblocks().

Signed-off-by: Dan Williams <dan.j.williams@intel.com>
Acked-by: Johannes Thumshirn <jthumshirn@suse.com>
---
 block/badblocks.c |    2 ++
 block/genhd.c     |    5 -----
 2 files changed, 2 insertions(+), 5 deletions(-)

--- a/block/badblocks.c
+++ b/block/badblocks.c
@@ -575,6 +575,8 @@ EXPORT_SYMBOL_GPL(devm_init_badblocks);
  */
 void badblocks_exit(struct badblocks *bb)
 {
+	if (!bb)
+		return;
 	kfree(bb->page);
 	bb->page = NULL;
 }
--- a/block/genhd.c
+++ b/block/genhd.c
@@ -670,11 +670,6 @@ void del_gendisk(struct gendisk *disk)
 	blk_unregister_queue(disk);
 	blk_unregister_region(disk_devt(disk), disk->minors);
 
-	if (disk->bb) {
-		badblocks_exit(disk->bb);
-		kfree(disk->bb);
-	}
-
 	part_stat_set_all(&disk->part0, 0);
 	disk->part0.stamp = 0;
 
